rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,640 @@
+  }();
+```
+
+## Op registration
+
+To support type specification, we will also provide a minimal Op registry and
+corresponding macro `REGISTER_KERNEL_FALLBACK_OP`.
+
+## Kernel implementation
+
+TensorFlow kernels inherit from the
+[OpKernel](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=82?q=opkernel)
+class and depend on two key classes:
+[OpKernelConstruction](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=256?q=opkernel)
+and
+[OpKernelContext](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=584?q=opkernel).
+We want to provide custom implementations of these two classes in terms of data
+we get from TFRT (for e.g. inputs, attributes).
+
+There are two main approaches to customize class implementations:
+
+*   Use inheritance and define common interfaces.
+*   Use templates.
+
+We ran multiple benchmarks to get an idea of the trade offs between inheritance
+and templating approaches. Key findings are summarized below:
+
+*   Time difference negligible for full model benchmarks.
+*   A simple scalar op benchmark with Kernel Fallback (runs scalar
+    multiplication, division, addition) was only 0.3% slower on mobile with
+    inheritance compared to templates.
+*   [basic\_ops\_benchmark](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/basic_ops_benchmark_test.cc?q=basic_ops_benchmark_test)
+    with inheritance is significantly slower: ~7% (median) or ~19% (mean)
+    (running on Linux). Note that this difference was measured *without* Kernel
+    Fallback. Adding inheritance would impact all existing TensorFlow kernels
+    even those that don't support Kernel Fallback.
+*   Binary size increase when using templates compared to inheritance is
+    estimated at 2.6% (based on adding `AddN` op).
+
+Right now, we are leaning towards using inheritance. Seems like time increase is
+only significant for running many scalar ops in a sequence - probably a rare use
+case in the real world.
+
+To use inheritance, we will define `OpKernelConstructionInterface` and
+`OpKernelContextInterface` interfaces. Ideally, these interfaces should be pure
+virtual. However, we will have one exception - templated `eigen_device` method
+that calls per-device pure-virtual implementations.
+
+We will then introduce `TFRTOpKernelConstruction` and `TFRTOpKernelContext`
+subclasses that implement `OpKernelConstructionInterface` and
+`OpKernelContextInterface` in terms of TFRT data structures. Example how
+`TFRTOpKernelConstruction` might look like:
+
+```cpp
+class TFRTOpKernelConstruction : public OpKernelConstructionInterface {
+ public:
+  explicit TFRTOpKernelConstruction(AttrMap attributes);
+  ~TFRTOpKernelConstruction() override {};
+
+  Status GetAttr(StringPiece attr_name, int32* value) const override;
```

Comment:
Consider a mechanism that allows delegated string interning to allow these attribute matches to ammortized by every invocation rather than multiplied by every invocation.

Reply:
We currently lookup attributes in TFRTs [OpAttrs](https://cs.opensource.google/tensorflow/runtime/+/master:include/tfrt/core_runtime/op_attrs.h;l=114?q=OpAttrs).
Depending how kernel fallback will be called from TFRT, `OpAttrs` might be passed to us as an argument and we save time by not creating a different data structure. That being said, these `OpAttrs` would only contain explicitly set attributes. We still need inferred attribute values.

So, may be a different data structure would make sense in the future and we can support string interning there.